sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+# Behavior for optional kernel features
+
+This design document describes the changes that are needed in DPC++ in order to
+conform to the SYCL 2020 specification regarding the behavior of applications
+that use optional kernel features.  An optional kernel feature is any feature
+that is supported by some devices and not by others.  For example, not all
+devices support 16-bit floating point operations, so the `sycl::half` data type
+is an optional kernel feature.  Some DPC++ extensions like AMX are also
+optional kernel features.
+
+The requirements for this design come mostly from the SYCL 2020 specification
+[section 5.7 "Optional kernel features"][1] but they also encompass the C++
+attribute `[[sycl::requires()]]` that is described in [section 5.8.1 "Kernel
+attributes"][2] and [section 5.8.2 "Device function attributes"][3].
+
+[1]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features>
+[2]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.attributes>
+[3]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_device_function_attributes>
+
+
+## Requirements
+
+There are several categories of requirements covered by this design:
+
+* The front-end compiler must issue a diagnostic in some cases when a kernel or
+  device function uses an optional feature.  However, the front-end compiler
+  must **not** generate a diagnostic in other cases.
+
+* The runtime must raise an exception when a kernel using optional features
+  is submitted to a device that does not support those features.  This
+  exception must be raised synchronously from the kernel invocation command
+  (e.g. `parallel_for()`).
```

Comment:
What happens if the user disables exceptions in the FE? (This is sort of a general question about SYCL moreso than this specific design document.)

Reply:
There are many places in the SYCL spec that mandate an exception to be thrown in various cases.  If the user disables exceptions, I suppose the implementation is no longer conformant.

There are places in the C++ runtime that mandate an exception also.  Doesn't that have the same issue if the user disables exceptions?

Reply:
The C++ standard doesn't admit that C++ is a thing without exceptions, so when users disable them in the compiler, they're technically compiling something C++-like but not really C++. But compilers generally disallow the user from throwing in that case (https://godbolt.org/z/TbPqKqv4c). If you can trick the compiler into throwing anyway, you get UB that may surprise you (https://godbolt.org/z/TbPqKqv4c) -- note that this comes from the ABI (https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html#cxx-aux).

I suppose in this case, the exception is thrown from the host side and expected to be caught from the host side, so the UB is what it is. I was more wondering whether there are cases where we throw from the device and the SYCL runtime expects to catch on the host side, and what we do when exceptions are disabled in that circumstance? But that's getting pretty far afield from this design doc, so don't feel obligated to answer/care. :-)

Reply:
> I was more wondering whether there are cases where we throw from the device and the SYCL runtime expects to catch on the host side

SYCL does not allow device code to throw C++ exceptions, so this situation cannot occur.  (SYCL disallows certain C++ features in device code, and exceptions are one such feature.)

If you don't mind, I'll mark this conversation resolved later today.  I'm trying to clean up the comment threads.

Reply:
Excellent to hear, thank you! I'll mark this as resolved now as there's nothing to be done for it.